File: sdk/go/provider/tempfastmail.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const tempfastmailBase = "https://tempfastmail.com"

type tempfastmailBox struct {
	Email string `json:"email"`
	UUID  string `json:"uuid"`
}

func TempfastmailGenerate() (*CreatedMailbox, error) {
	req, err := http.NewRequest(http.MethodPost, tempfastmailBase+"/api/email-box", nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "Mozilla/5.0")
	resp, err := HTTPClient().Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("tempfastmail create: http %d", resp.StatusCode)
	}
	var data tempfastmailBox
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	if !strings.Contains(data.Email, "@") || data.UUID == "" {
		return nil, fmt.Errorf("tempfastmail: invalid create response")
	}
	return &CreatedMailbox{
		Channel: "tempfastmail",
		Email:   strings.TrimSpace(data.Email),
		Token:   strings.TrimSpace(data.UUID),
	}, nil
}

func tempfastmailFlatten(raw map[string]any, recipient string) map[string]any {
	to := raw["real_to"]
	if s, ok := to.(string); !ok || strings.TrimSpace(s) == "" {
		to = recipient
	}
	return map[string]any{
		"id":          raw["uuid"],
		"from":        raw["from"],
		"to":          to,
		"subject":     raw["subject"],
		"html":        raw["html"],
		"date":        raw["received_at"],
		"attachments": raw["attachments"],
	}
}

func tempfastmailJSON(path string, out any) error {
	req, err := http.NewRequest(http.MethodGet, tempfastmailBase+path, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "Mozilla/5.0")
	resp, err := HTTPClient().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("tempfastmail: http %d", resp.StatusCode)
	}
	return json.Unmarshal(body, out)
}

func TempfastmailGetEmails(token string, email string) ([]NormEmail, error) {
	uuid := strings.TrimSpace(token)
	address := strings.TrimSpace(email)
	if uuid == "" {
		return nil, fmt.Errorf("tempfastmail: empty token")
	}
	if address == "" {
		return nil, fmt.Errorf("tempfastmail: empty email")
	}
	var rows []map[string]any
	if err := tempfastmailJSON(fmt.Sprintf("/api/email-box/%s/emails", url.PathEscape(uuid)), &rows); err != nil {
		return nil, err
	}
	out := make([]NormEmail, 0, len(rows))
	for _, row := range rows {
		raw := row
		if messageID, _ := row["uuid"].(string); strings.TrimSpace(messageID) != "" {
			var detail map[string]any
			if err := tempfastmailJSON(fmt.Sprintf("/api/email-box/%s/email/%s", url.PathEscape(uuid), url.PathEscape(messageID)), &detail); err == nil {
				raw = detail
			}
		}
		out = append(out, NormalizeMap(tempfastmailFlatten(raw, address), address))
	}
	return out, nil
}

File: sdk/npm/src/providers/tempfastmail.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "tempfastmail";
const BASE_URL = "https://tempfastmail.com";

interface EmailBoxResponse {
  email?: string;
  uuid?: string;
}

function flattenMessage(
  raw: Record<string, unknown>,
  recipient: string,
): Record<string, unknown> {
  return {
    id: raw.uuid,
    from: raw.from,
    to: raw.real_to || recipient,
    subject: raw.subject,
    html: raw.html,
    date: raw.received_at,
    attachments: raw.attachments,
  };
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const response = await fetchWithTimeout(`${BASE_URL}/api/email-box`, {
    method: "POST",
    headers: {
      Accept: "application/json",
      "User-Agent": "Mozilla/5.0",
    },
  });
  if (!response.ok) {
    throw new Error(`tempfastmail create http ${response.status}`);
  }
  const data = (await response.json()) as EmailBoxResponse;
  const email = String(data.email || "").trim();
  const uuid = String(data.uuid || "").trim();
  if (!email || !email.includes("@") || !uuid) {
    throw new Error("tempfastmail: invalid create response");
  }
  return {
    channel: CHANNEL,
    email,
    token: uuid,
  };
}

export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  const uuid = String(token || "").trim();
  const address = String(email || "").trim();
  if (!uuid) throw new Error("tempfastmail: empty token");
  if (!address) throw new Error("tempfastmail: empty email");

  const listResponse = await fetchWithTimeout(
    `${BASE_URL}/api/email-box/${encodeURIComponent(uuid)}/emails`,
    {
      headers: {
        Accept: "application/json",
        "User-Agent": "Mozilla/5.0",
      },
    },
  );
  if (!listResponse.ok) {
    throw new Error(`tempfastmail list http ${listResponse.status}`);
  }
  const rows = (await listResponse.json()) as Array<Record<string, unknown>>;
  if (!Array.isArray(rows)) return [];

  const emails: Email[] = [];
  for (const row of rows) {
    const messageId = String(row.uuid || "").trim();
    let raw = row;
    if (messageId) {
      const detailResponse = await fetchWithTimeout(
        `${BASE_URL}/api/email-box/${encodeURIComponent(uuid)}/email/${encodeURIComponent(messageId)}`,
        {
          headers: {
            Accept: "application/json",
            "User-Agent": "Mozilla/5.0",
          },
        },
      );
      if (detailResponse.ok) {
        raw = (await detailResponse.json()) as Record<string, unknown>;
      }
    }
    emails.push(normalizeEmail(flattenMessage(raw, address), address));
  }
  return emails;
}
